### PR TITLE
bind -> rbind.fill; getMetadata; GeoJSON method

### PR DESCRIPTION
**Pull request by dmpe**
## Functions to speed up the row binding
So, to begin with, there are 4 (to me) known functions which can be used to bind rows together. These are: 

- `rbind` from `base`
This is is known to be slow with *large* datasets. For example, see this: http://rpubs.com/wush978/6302

- `rbind.fill` from `plyr`
A faster version of `rbind`, implemented in R. And the only other function which can be used for our purpose. 

- `bind_rows` from `dplyr`: This is a faster (C-implemented) version of `rbind.fill` from `plyr`. The only significant change for us is the fact that it is far more stricter in terms of types which can be permitted in the column. While plyr "doesn't seem to care much" and will "just do it", `dplyr` will check it and throw an error if there is e.g. an integer and character in the same column. There is already a bug filled, which has been closed https://github.com/hadley/dplyr/issues/1162 with hadley saying 
`
I'm currently happy with the behaviour of bind_rows(), and if you'd prefer the rbindlist() behaviour, you're free to use that. 
`
Thus, as we cannot really on it (and from my own experience working with dplyr), this package would be out of touch. 

- `rbindlist` from `data.table`:
`
I'm hesitant of data.table. I'm not a huge fan of their slightly off-kilter data frame. data.table is fast, but I worry they just aren't going to get a critical mass of users to make it well-known to many R programmers.
`
Yes, I fully agree on this and this is also a reason why i have decided not to apply it here. ~~If we have decided to do it, then we would make it in part backwards incompatible (and would be required to run a marketing campaign to our users :) )~~

### Tests
I have commited the benchmarking file (Rmd) here without appropriate functions because otherwise it would be a huge mess. Here is a gist for them: https://gist.github.com/dmpe/5aec87f0c7a5ae2115ca

- [x] [Benchmark rbind @ google ](https://www.google.cz/search?q=benchmark+rbind)
 - [x] A very good one from 2014: http://www.exegetic.biz/blog/2014/06/concatenating-a-list-of-data-frames/
 - [x] With our sample data: http://rpubs.com/F789GH/BenchmarkRowsBinding
   - While with the first one, I cannot apply `dplyr` because it will print an error, with the second one there is no such a problem. This just confirms the issue with `dplyr` I have described above. Again, with the third one below, `dplyr` will fail again. 
    - Looking at mean in seconds: 1. `datatable`; 2. `plyr`; 3 . `base`; 4. `dplyr` (? what ?)
    - Looking at median in seconds: 1. `datatable`; 2. `dplyr`; 3 . `plyr`; 4. `base` (This already makes sense)
    - **Summary**: Given that `dplyr` is complicated and `data.table` is a no-go, the only options is here either to use `plyr` or stick with `base` `rbind`. Due to the small (and with larger datasets bigger) speed up, I decided to choose `plyr`. But this can be - of course - further discussed. BTW, I also think and actually saw that it will be largely a dataset dependent, i.e. with some smaller datasets `rbind` would be faster, while with some other `plyr` would be better. 

**Third one**:
```
big <- microbenchmark(
  read.socrataRBIND("https://data.ny.gov/resource/hrvs-fxs2.csv"),
  read.socrataPLYR("https://data.ny.gov/resource/hrvs-fxs2.csv"),
  read.socrataDPLYR("https://data.ny.gov/resource/hrvs-fxs2.csv"),
  read.socrataDATATABLE("https://data.ny.gov/resource/hrvs-fxs2.csv"),
  times = 3L,
  unit = "s"
)
big
Unit: seconds
                                                                expr      min       lq     mean   median       uq      max neval cld
     read.socrataRBIND("https://data.ny.gov/resource/hrvs-fxs2.csv") 17.24245 17.43700 17.80954 17.63156 18.09309 18.55463     3   a
      read.socrataPLYR("https://data.ny.gov/resource/hrvs-fxs2.csv") 17.39023 17.41183 17.43587 17.43343 17.45869 17.48395     3   a
 read.socrataDATATABLE("https://data.ny.gov/resource/hrvs-fxs2.csv") 16.90609 16.93185 17.24587 16.95761 17.41576 17.87392     3   a
```
This will fail at dplyr again; without it: 1. `datatable`; 2. `plyr`; 3. `base `

## Detecting file output (geojson, csv, json)

`Excellent! I wonder, though, if this should be a different method, we detect it using the extension, or use an option within read.socrata. For instance, here are alternative scenarios:`

Yes, I though about that when creating this patch set.

The reason why I have decided for splitting GEO method was, that the method `read.socrata` would be otherwise really huge.
Given your `geojson` choice, the method `geojson_read` from `geojsonio` package offers several arguments such as `method`,`parse`, `what` and these must be accessible for the user. Thus you would do something like 

`
read.socrata <- function(url = NULL, app_token = NULL, limit = 50000, domain = NULL, fourByFour = NULL, query = NULL, offset = 0, output = c("csv", "json", "geojson"), ...) {}
`
where ... would be used for `geojson_read` arguments.
OR 
`
read.socrata <- function(url = NULL, app_token = NULL, limit = 50000, domain = NULL, fourByFour = NULL, query = NULL, offset = 0, output = c("csv", "json", "geojson"), method = "something", parse = "something", what="something") {}
`
However, the last 3 will apply only to the "geojson" choice. One would document that in the documentation, but just looking at such function I would be confused. 

The current read.socrataGEO method is small and very clear to the user as it deals only with the GEO stuff. While the other one, is used for CSV & JSON stuff exclusively. 

**But**, look, I agree that this is largely a personal preference and nothing stops us in doing what Tom has suggested. This is also a reason why I wanted some feedback on my changes. 

### Changed behaviour
- I have put application token into the header of `httr::GET()` request in the `errorHandling()` method, so we can speare dealing with `$$token=` somewhere in the code. Thus, tests have been edited for this change

- I have switched to json only (kind of) and thus added new validation logic.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -19,6 +19,7 @@ r_github_packages:
   - jeroenooms/curl
   - klutometis/roxygen
   - jimhester/covr
+  - yihui/mime
   - ropensci/geojsonio
```
Comment (dmpe) [inline .travis.yml]: We now also require `mime` version >= 4.0 which has been published to CRAN
https://cran.rstudio.com/web/packages/mime/index.html

File: DESCRIPTION [hunk 2]
```diff
@@ -5,27 +5,28 @@ Description: Provides easier interaction with
...
+    plyr (>= 1.8.3)
 Suggests: 
     testthat (>= 0.10.0),
-    roxygen2 (>= 4.1.0),
-    knitr (>= 1.10.5), 
-    leaflet (>= 1.0.0), 
-    geojsonio (>= 0.1.0)
+    roxygen2 (>= 4.1.1),
+    knitr (>= 1.11), 
+    leaflet (>= 1.0.0)
```
Comment (dmpe) [inline DESCRIPTION]: used in the vignette with examples

File: R/returnData.R [hunk 3]
```diff
@@ -32,6 +16,7 @@ checkResponse <- function(url = "") {
 #'
 #' @author Hugh J. Devlin \email{Hugh.Devlin@@cityofchicago.org}
 #' @importFrom httr content
+#' @importFrom geojsonio geojson_read
```
Comment (dmpe) [inline R/returnData.R]: TODO: myself

File: R/returnData.R [hunk 4]
```diff
@@ -59,77 +44,148 @@ getContentAsDataFrame <- function(response) {
...
   
+  rowCount <- as.numeric(getMetadata(cleanQuest(validUrl))[1])
+  
   ## More to come? Loop over pages implicitly
-  while (nrow(page) > 0) { 
-    query_url <- paste0(validUrl, ifelse(is.null(parsedUrl$query), '?', "&"), '$offset=', nrow(result))
-    response <- checkResponse(query_url)
+  while (nrow(results) < rowCount) { 
+    query_url <- paste0(validUrl, ifelse(is.null(parsedUrl$query), "?", "&"), "$offset=", nrow(results), "&$limit=", limit)
+    response <- errorHandling(query_url)
```
Comment (dmpe) [inline R/returnData.R]: TODO: myself: -> order when downloading

File: R/returnData.R [hunk 5]
```diff
@@ -1,34 +1,46 @@
 # An interface to data hosted online in Socrata data repositories
 # This is the main file which uses other functions to download data from a Socrata repositories
 #
-# Author: Hugh J. Devlin, Ph. D. 2013-08-28
+# Author: Hugh J. Devlin, Ph.D et al.
```
Comment (dmpe) [inline R/returnData.R]: to be fixed later

**Comment by dmpe** [comment]
With this PR, following issues can be marked as closed:
https://github.com/Chicago/RSocrata/issues/43
https://github.com/Chicago/RSocrata/issues/19
https://github.com/Chicago/RSocrata/issues/27
https://github.com/Chicago/RSocrata/issues/33

File: R/returnData.R [hunk 6]
```diff
@@ -1,34 +1,46 @@
...
+    data.frame(t(sapply(con, unlist)), stringsAsFactors = FALSE)
+  } else {
+    all_cols <- names(con[[which(sapply(con, length) == max(sapply(con, length)))[1]]])
+    con <- lapply(con, function(x) {
+      r <- c(x, sapply(all_cols[!all_cols %in% names(x)], function(xx) NA, simplify = FALSE))
+      r[all_cols]
+    })
+    data.frame(t(sapply(con, unlist)), stringsAsFactors = FALSE)
+  }
 }
```
Comment (dmpe) [inline R/returnData.R]: this now fixes the issue with uneven JSON columns.

**Comment by dmpe** [comment]
Hi @tomschenkjr @geneorama , 
could you please take a look on these preliminary changes. Thank you

File: R/validateURL.R [hunk 7]
```diff
@@ -1,62 +1,81 @@
...
+  if(!isFourByFour(cleanDot(fourByFour))) {
+    stop(fourByFour, " is not a valid Socrata dataset unique identifier.")
+  }
+  
+  if( parsedUrl$scheme == "http") {
+    parsedUrl$scheme <- "https"
+  }
+  
+  # http://stackoverflow.com/a/7964098
+  # For TRANSITION: it will be easier for users of CSV to translate to JSON by warning them.
```
Comment (dmpe) [inline R/validateURL.R]: transit

**Comment by tomschenkjr** [comment]
@dmpe - Quite a few notes to get through.

At a high-level, I think it's important to remember two key goals of the package. Both of which basically sum-up to making it "easy" to work with Socrata URLs:

1. Allow users to copy/paste either a valid SoDA URI or "human-readable" URL without having to handle the web call itself (e.g., paging) to put it into a convenient storage type
2. Have the package translate an handle data types behind-the-scenes to be convenient in R (e.g., dates)

There are three primary interactions with RSocrata:

A. Use a valid SoDA API asking for a JSON file: e.g., https://data.example.com/resource/four-four.json
B. Use a valid SoDA API asking for a CSV file: e.g., https://data.example.com/resource/four-four.csv
C. Use a human readable API e.g., https://data.example.com/Human-Readable-Name/four-four
D. Proposed: Use a human readable GeoJSON output.

Finally, in the spirit of [semantic versioning](www.semver.org), we should be mindful of introducing anything that is not backward compatible for the end-user. Everything "behind-the-scenes" can be reworked more frequently to find the better solution. Breaking backward compatibility should be delayed if not entirely avoided.

The above items should be our central tenants. As long as these are met, then I'm good with any modification.

So, keeping that in mind with the comment above, here are my thoughts:
* I don't mind which method is used for binding rows, as long as it works. `data.table` seems to violate (1) given it's odd data storage. Otherwise, whatever is fastest works.
* For file output, we should avoid overthinking this. Basically, it should meet criteria (A), (B), and (C). Behind-the-scenes doesn't matter. So, I think it's overcomplicating by asking the user. We should just make the three primary interactions work. So, we should remove user-options on this and optimize the end-user experience by making our own choice.
* For `read.socrataGeo`, your explanation makes sense. However, let me ask a question-- does the user really need to access `parse`, `method` and `what`, or is this something we can choose? I ask because, since this is Socrata, all of those options may be the same regardless of the specific dataset.
* I think the switch to JSON only is not bad because it's a "backend" work. However, *I think the biggest issue is the lack of backwards compatibility with human-readable URLs.* This breaks the spirit of semantic versioning.

**Comment by dmpe** [comment]
Hi Tom, 
Thanks for the reply. See my answers below. 

> I don't mind which method is used for binding rows, as long as it works. data.table seems to violate (1) given it's odd data storage. Otherwise, whatever is fastest works.

Plyr

> For file output, we should avoid overthinking this. Basically, it should meet criteria (A), (B), and (C). Behind-the-scenes doesn't matter. So, I think it's overcomplicating by asking the user. We should just make the three primary interactions work. So, we should remove user-options on this and optimize the end-user experience by making our own choice.

> I think the switch to JSON only is not bad because it's a "backend" work. 

Yep, no option for the user. Everything will be done in JSON. 

> For read.socrataGeo, your explanation makes sense. However, let me ask a question-- does the user really need to access parse, method and what, or is this something we can choose? I ask because, since this is Socrata, all of those options may be the same regardless of the specific dataset.

We can certainly have this: 
`results <- geojsonio::geojson_read(url, method = "local", what = "list", parse = FALSE)`. Then, the method would have only two arguments, that is the URL and `...`. 

> However, I think the biggest issue is the lack of backwards compatibility with human-readable URLs. This breaks the spirit of semantic versioning.

Fail. I wrote something where text and code are saying two different things. In fact, the current changes are backwards compatible! What it does is that it takes a `http://soda.demo.socrata.com/dataset/USGS-Earthquake-Reports/4334-bgaj.csv` and replaces it with a SODA one. 

Sorry for confusion! I deleted that previous message above.

In case, you are going to download data using human-readable URL, with a csv suffix you will get this:
```
BEWARE: Your suffix is no longer supported. Thus, we will automatically replace it with JSON.
BEWARE: RSocrata no longer supports human readable URLs, thus we will convert it to the SODA 
            ones (with resource/xxxx-xxxx.json) instead.
```

**Comment by dmpe** [comment]
Hi, @tomschenkjr Any update on this ? I believe i have addressed your points. But if not, please let me know. 
Thanks.

**Comment by tomschenkjr** [comment]
Looks good. Pulling into the dev branch.

**Comment by dmpe** [comment]
Thank you !

**Comment by tomschenkjr** [comment]
FYI, I have a small project that I'm running the dev version and hitting some issues that I'll be posting on. In the meantime, here is the code I'm running:

```r
ls_chicago <- ls.socrata("https://data.cityofchicago.org")
four_by_four <- substr(ls_chicago$identifier, 42, 50)

read.socrata(url = "http://data.cityofchicago.org", 
             fourByFour = four_by_four[1])
```
Resulting in:
```
Error in validateUrl(url) : 
   is not a valid Socrata dataset unique identifier.
```

Also, the unit testing seemed to fail, but can resolve those shortly.

**Comment by dmpe** [comment]
Thanks, I saw that too. 
I will take a look into the first thing and you can take the second one.

**Comment by dmpe** [comment]
Ok, for one, you need to replace 
```
read.socrata(url = "http://data.cityofchicago.org", 
             fourByFour = four_by_four[1])
```
with 
```
 read.socrata(domain = "http://data.cityofchicago.org", 
               fourByFour = four_by_four[1])
```
Did it work always ? Because I don't know that we can even do so. URL parameters always - I believe - consisted of the whole, complete URL

**Comment by tomschenkjr** [comment]
Got it, that's correct. Sorry, my transcription error. 

The error I had meant to post was
```
Error in curl::curl_fetch_memory(url, handle = handle) : 
  Couldn't resolve host name
```

**Comment by dmpe** [comment]
I got this error too. BUT at random, non reproducable. I have no idea how to deal with it. Plus, what does that even mean, because `asd <- httr::GET("https://data.cityofchicago.org/resource/kc9i-wq85.json", httr::add_headers("X-App-Token" = NULL))` is for me at least a non issue ?

**Comment by dmpe** [comment]
Plus it is case by case basis. Some work without the error, while others not. 

I will soon provide you a link (not a PR), which will get some things better (but not this)

**Comment by tomschenkjr** [comment]
Ok, [this](https://github.com/Chicago/RSocrata/blob/b8df56daf10da89c64b8dfe681b9c39ed42d7bca/R/returnData.R#L111) seems to be the offending line.

I'm submitting a couple of corrections to dev after I check out a few more things and write a couple of tests to try to catch these errors in the future.

I'm also going to remove the warnings being thrown on human-readable URL support. Since it's not being deprecated, and it's a backend adjustment that doesn't directly impact the code, just need to note it in release notes.

**Comment by tomschenkjr** [comment]
Pushed changes to the repo. Took awhile to get this into build into a package, so need to keep that in order.

Having trouble building the examples, so need to insert more `\dontrun{}` around the examples. I opened #60

I also had to remove the vignettes to build the package. Opened #59 to note their addition again.

The new geojson feature will be great. Will need some tweaks but will be able to get into 1.7.

**Comment by dmpe** [comment]
@tomschenkjr ad Problem with build URL: Please, take also a look on https://cran.r-project.org/web/packages/urltools/index.html as I am more and more now with the university stuff and have less and less time to boot VM and work on Rsocrata. 
If we could replace httr 2 methods I would be highly in favor.

**Comment by dmpe** [comment]
PS: https://github.com/Chicago/RSocrata/blob/dev/R/returnData.R#L102
This `domain = "http://data.cityofchicago.org"` should be also now documented that `domain = "http://data.cityofchicago.org/"` (with / at the end) will not work properly.

**Pull request merged**